File: impl/sdk/py/src/resonate/durable.py
"""Runtime representation of user functions registered as durable.

Python annotations are optional, so ``def fn(ctx, x):`` carries nothing that
says ``ctx`` is a :class:`Context`. The SDK therefore adopts one convention:
**every** durable function -- workflow or leaf -- takes a :class:`Context` as
its first positional argument. The runtime strips that first parameter and
injects the ``Context`` on each call; it never inspects annotations to decide
*shape*.

Annotations drive exactly one thing: **symmetric serialize/deserialize**. For
the root / dispatched path a call's arguments are JSON-encoded into the durable
promise's ``param`` and, on every (re-)invocation, coerced back to their declared
parameter types (:meth:`DurableFunction.invoke` with ``coerce_args=True``); a
recovered return value is likewise coerced back to the declared return type
(:meth:`DurableFunction.coerce_result`). Both run through the one primitive
:meth:`DurableFunction._coerce_value`, the same type-shaping step the top-level
:class:`~resonate.handle.ResonateHandle` applies. A missing /
``Any`` / unresolved annotation is a pass-through, so an untyped function behaves
like ``rpc`` (raw builtins, the caller's responsibility).

A local ``ctx.run`` child is the exception: its arguments are never serialized
into the param (nothing reads it back -- see :meth:`Context.run`), so
:meth:`DurableFunction.invoke` is called with ``coerce_args=False`` and the
in-memory objects reach the function verbatim. That is what lets ``ctx.run``
accept non-serializable arguments. The return value still round-trips and is
coerced, so it must remain serializable.
"""

from __future__ import annotations

import inspect
from typing import TYPE_CHECKING, Any, Concatenate

import msgspec

from resonate.error import ApplicationError, SerializationError
from resonate.types import Args

if TYPE_CHECKING:
    from collections.abc import Callable

    from resonate.context import Context


class DurableFunction:
    """A user function adapted for durable execution.

    Built once per function (at registration / ``ctx.run`` time). It remembers
    how to re-bind serialized arguments on every (re-)invocation so a
    suspended-then-resumed call reconstructs the same Python call it was first
    dispatched with. The :class:`Context` is always injected as the first
    positional argument; only the *user* arguments round-trip through the durable
    promise's ``param`` field.
    """

    def __init__(self, fn: Callable[Concatenate[Context, ...], Any]) -> None:
        if not callable(fn):
            msg = f"expected a callable, got {type(fn).__name__}"
            raise ApplicationError(msg)

        name = getattr(fn, "__name__", "unknown")
        sig = inspect.signature(fn)
        params = list(sig.parameters.values())
        if not params:
            msg = (
                f"{name}: durable function must accept a Context as its first argument"
            )
            raise ApplicationError(msg)

        # ``ctx`` is injected positionally (``self._fn(ctx, ...)``) and stripped by
        # position (``params[1:]``), so the first parameter must be a plain
        # positional one. A ``*args`` / ``**kwargs`` / keyword-only leading
        # parameter cannot hold it -- and a bare ``(*args, **kwargs)`` is the
        # tell-tale of a decorator applied without :func:`functools.wraps`, which
        # erases the real signature (and with it arity validation and coercion).
        # Reject it with a clear message rather than failing confusingly later.
        if params[0].kind not in (
            inspect.Parameter.POSITIONAL_ONLY,
            inspect.Parameter.POSITIONAL_OR_KEYWORD,
        ):
            msg = (
                f"{name}: durable function must accept a Context as its first "
                f"positional argument, but its first parameter is "
                f"{params[0].kind.description!r} (if this is a decorated function, "
                f"apply functools.wraps so its real signature is visible)"
            )
            raise ApplicationError(msg)

        # Resolve string annotations (``from __future__ import annotations`` turns
        # them all into strings) tolerantly and per-annotation: a name that is not
        # importable at runtime -- e.g. a ``Context`` under ``TYPE_CHECKING`` -- is
        # left as a string and treated as pass-through, without disabling coercion
        # of its resolvable siblings. Only the *user* parameters and the return are
        # resolved; the stripped ``ctx`` slot is never coerced, so its (often
        # unresolvable) annotation is ignored.
        globalns = _func_globals(fn)

        self._fn = fn
        #: Signature of the *user* arguments (leading ``ctx`` removed, annotations
        #: resolved). The runtime owns the ``ctx`` slot and never exposes it.
        self._user_sig = sig.replace(
            parameters=[
                p.replace(annotation=_resolve_annotation(p.annotation, globalns))
                for p in params[1:]
            ]
        )
        #: Resolved return annotation, used by :meth:`coerce_result`.
        self._return_annotation = _resolve_annotation(sig.return_annotation, globalns)
        #: Source name (child context ``func_name`` + error messages). The registry
        #: key is supplied separately at register time and may differ.
        self.name = name

    def pack_args(self, *args: Any, **kwargs: Any) -> Args:
        """Validate a call and pack it into a serializable :class:`Args`.

        Called at dispatch (``ctx.run``) time. :class:`Args` is the single slot
        able to round-trip Python ``*args`` / ``**kwargs`` through the durable
        promise's one ``param`` field; an empty call yields ``Args()``. The
        injected ``Context`` is never part of the payload. Raises
        :class:`ApplicationError` if the call does not match the signature.
        """
        try:
            self._user_sig.bind(*args, **kwargs)
        except TypeError as exc:
            msg = f"{self.name}: {exc}"
            raise ApplicationError(msg) from exc

        return Args(args=args, kwargs=kwargs)

    async def invoke(
        self, ctx: Context, packed: Args, *, coerce_args: bool = True
    ) -> Any:
        """Re-bind ``packed`` to the signature and execute the function.

        ``packed`` is the :class:`Args` :meth:`pack_args` produced (a local child)
        or the :class:`~resonate.types.TaskData` Core decoded from the root promise
        param -- ``TaskData`` is an ``Args``, so both arrive through the same slot.

        When ``coerce_args`` is ``True`` (the root / dispatched path) arguments are
        coerced to their declared types: that path's args arrive as JSON builtins
        decoded from the persisted param, so they must be reshaped, and msgspec
        strictness applies (e.g. ``True`` is rejected for an ``int`` parameter, an
        ``int`` is widened to a declared ``float``).

        A local ``ctx.run`` child passes ``coerce_args=False``: its arguments are
        never serialized into the promise param (see :meth:`Context.run`), so the
        in-memory objects reach the function verbatim -- nothing to reshape, and
        no annotation can reject them. This is what lets ``ctx.run`` accept
        non-serializable arguments. Symmetry with replay is preserved because the
        parent re-derives the *same* in-memory objects on every run, so neither
        path round-trips. ``bind`` + ``apply_defaults`` still run, so arity is
        validated and defaults are filled regardless.

        ``ctx`` is injected as the first positional argument; sync and ``async``
        functions are both supported.
        """
        try:
            bound = self._user_sig.bind(*packed.args, **packed.kwargs)
        except TypeError as exc:
            msg = f"{self.name}: {exc}"
            raise ApplicationError(msg) from exc

        # Snapshot which arguments were actually supplied *before* apply_defaults
        # fills the rest: only supplied values crossed the serialization boundary
        # and need coercion. A default is a live object (possibly a sentinel whose
        # type deliberately violates its annotation), so it is left untouched.
        provided = set(bound.arguments)
        bound.apply_defaults()
        if coerce_args:
            self._coerce(bound, provided)

        result = self._fn(ctx, *bound.args, **bound.kwargs)
        if inspect.isawaitable(result):
            return await result
        return result

    def coerce_result(self, value: Any) -> Any:
        """Reshape a recovered return value into the declared return type.

        The symmetric counterpart of argument coercion. ``ctx.run`` calls this on
        both paths -- the recovery short-circuit (the settled child value comes
        back as JSON builtins) and the live path (the raw object the function
        returned) -- so the parent observes the same in-memory value either way.

        .. warning::
            A pass-through annotation (unannotated / ``Any`` / ``None`` /
            unresolved) reshapes nothing, so a non-JSON-native value round-trips
            faithfully only when the annotation can rebuild it: an untyped (or
            ``Any``) ``datetime`` / ``set`` return yields the live object on the
            first run but the decoded builtin (``str`` / ``list``) on recovery.
            Annotate the return type to keep the two paths symmetric.
        """
        return self._coerce_value(value, self._return_annotation)

    @property
    def return_type(self) -> Any:
        """The declared return type, as a value safe to hand to ``msgspec.convert``.

        The type-shaped view of the resolved return annotation, for a caller that
        decodes a settled value against a *type* rather than calling
        :meth:`coerce_result` -- namely the top-level
        :class:`~resonate.handle.ResonateHandle`, whose generic ``T`` is erased at
        runtime. A pass-through annotation collapses to ``Any`` so
        ``msgspec.convert(value, return_type)`` is the same no-op
        :meth:`coerce_result` performs for those cases, keeping the top-level
        decode and the ``ctx.run`` recovery decode in lock-step (and resolving the
        annotation through the one tolerant path that also handles callable
        instances, unlike a hand-rolled re-resolution).
        """
        if _is_passthrough_annotation(self._return_annotation):
            return Any
        return self._return_annotation

    def _coerce(self, bound: inspect.BoundArguments, provided: set[str]) -> None:
        """Coerce each *supplied* argument to its declared type, in place.

        ``*args`` / ``**kwargs`` coerce element-wise (their annotation is the
        element type); every other parameter coerces its single value. Defaults
        filled by ``apply_defaults`` are not in ``provided`` and are skipped.
        """
        for name in provided:
            param = self._user_sig.parameters[name]
            annotation = param.annotation
            value = bound.arguments[name]
            if param.kind is inspect.Parameter.VAR_POSITIONAL:
                bound.arguments[name] = tuple(
                    self._coerce_value(v, annotation) for v in value
                )
            elif param.kind is inspect.Parameter.VAR_KEYWORD:
                bound.arguments[name] = {
                    k: self._coerce_value(v, annotation) for k, v in value.items()
                }
            else:
                bound.arguments[name] = self._coerce_value(value, annotation)

    def _coerce_value(self, value: Any, annotation: Any) -> Any:
        """Reshape one value to ``annotation`` via msgspec, or pass it through.

        The single coercion primitive shared by argument coercion (:meth:`_coerce`)
        and return coercion (:meth:`coerce_result`), so the serialize/deserialize
        pair cannot drift. An annotation that carries nothing to reshape (see
        :func:`_is_passthrough_annotation`) leaves the value untouched.
        """
        if _is_passthrough_annotation(annotation):
            return value

        try:
            return msgspec.convert(value, annotation)
        except (TypeError, ValueError, msgspec.MsgspecError) as exc:
            error = SerializationError(exc)
            error.add_note(f"while binding arguments for {self.name}")
            raise error from exc


def _is_passthrough_annotation(annotation: Any) -> bool:
    """Whether ``annotation`` carries no information to reshape a value.

    Shared by :meth:`DurableFunction._coerce_value` (argument + return coercion)
    and :meth:`DurableFunction.return_type` (the top-level decode type), so every
    view of "is this typed?" agrees. Four cases pass through untouched:

    * ``empty``  -- the parameter / return is genuinely unannotated.
    * ``Any``    -- explicitly opted out of typing.
    * ``None``   -- ``-> None`` / ``x: None``; the value is already ``None``.
    * a ``str``  -- a forward ref :func:`_resolve_annotation` could not evaluate.
    """
    return (
        annotation is inspect.Parameter.empty
        or annotation is Any
        or annotation is None
        or isinstance(annotation, str)
    )


def _func_globals(fn: Callable[Concatenate[Context, ...], Any]) -> dict[str, Any]:
    """Best-effort global namespace for resolving ``fn``'s string annotations.

    A plain function or bound method exposes ``__globals__`` directly. A *callable
    instance* does not -- its annotations live on ``type(fn).__call__`` -- so fall
    back to that callable's globals. Without this fallback an instance's
    annotations resolve against an empty namespace, so every non-builtin
    annotation silently stays an unresolved string and coercion is skipped,
    handing the function a raw ``dict`` / ``list`` instead of the declared type.
    """
    globalns = getattr(fn, "__globals__", None)
    if globalns is not None:
        return globalns
    return getattr(type(fn).__call__, "__globals__", {})


def _resolve_annotation(annotation: Any, globalns: dict[str, Any]) -> Any:
    """Evaluate one string annotation, leaving it as-is if it cannot resolve.

    Resolution is per-annotation and best-effort: ``inspect.signature(eval_str=True)``
    is all-or-nothing (a single unresolvable name aborts the whole signature),
    which would silently disable coercion of every sibling whenever a function
    follows the standard ``TYPE_CHECKING``-only ``Context`` import. Here each
    annotation is evaluated alone; one that fails yields its raw string, which
    :meth:`DurableFunction._coerce_value` then treats as pass-through.

    Any evaluation failure is swallowed, not just ``NameError`` -- an annotation
    string is arbitrary Python that ``eval_str=True`` already runs, so a broad
    ``except`` adds no new risk, and a value that cannot be reshaped is left
    un-coerced rather than aborting registration of an otherwise-valid function.
    """
    if not isinstance(annotation, str):
        return annotation
    holder = type("_", (), {"__annotations__": {"a": annotation}})
    try:
        return inspect.get_annotations(holder, globals=globalns, eval_str=True)["a"]
    except Exception:
        return annotation

File: impl/sdk/java/src/main/java/io/resonatehq/resonate/Durable.java
package io.resonatehq.resonate;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.resonatehq.resonate.Errors.ApplicationError;
import io.resonatehq.resonate.Errors.SerializationError;
import io.resonatehq.resonate.Types.Args;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Runtime representation of a user function registered as durable, mirroring {@code
 * resonate.durable.DurableFunction} from the Python SDK.
 *
 * <p><b>The {@code inspect} → reflection mapping.</b> Python introspects a callable with {@code
 * inspect.signature}: parameter kinds, per-parameter type annotations, and the return annotation
 * drive both the ctx-first convention and the symmetric serialize/deserialize coercion. Java's
 * analogue is {@link java.lang.reflect.Method} — a {@code Method} (unlike a bare lambda) exposes its
 * parameter <em>types</em>, generic element types, varargs-ness, and return type at runtime, so a
 * durable function here is supplied as a reflected {@code Method} (plus an optional target instance
 * for instance methods).
 *
 * <p><b>Convention.</b> Every durable function takes a {@code Context} as its first positional
 * argument; the runtime strips that slot by position (it never inspects the slot's declared type)
 * and injects the context on each call. Only the <em>user</em> parameters round-trip through the
 * durable promise's {@code param} field as an {@link Args} slot.
 *
 * <p><b>Coercion.</b> On the dispatched/recovery path arguments arrive as JSON builtins (a {@code
 * Map} for a record, a {@code List} for an array) and must be reshaped to each parameter's declared
 * (generic) type via Jackson — the analogue of {@code msgspec.convert}. A parameter typed {@link
 * Object} carries nothing to reshape and passes through (the {@code Any} analogue), exactly as an
 * unannotated Python parameter does. The recovered return value is reshaped the same way through
 * {@link #coerceResult}, so the live and replay paths stay symmetric.
 *
 * <p><b>Divergences from Python (no Java analogue).</b> Java has no keyword arguments, keyword-only
 * or positional-only parameters, or default parameter values, so the kwargs slot is carried for
 * wire compatibility but never bound by name, and {@code apply_defaults} has no counterpart. Java
 * methods always carry concrete parameter types, so the "unresolved/forward-ref annotation"
 * tolerance and the {@code __globals__} fallback are likewise moot.
 */
public final class Durable {

    // Jackson is this repo's msgspec analogue (see Types / Codec); convertValue == msgspec.convert.
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final Object target; // null for a static method
    private final Method fn;
    private final String name;

    // The user parameters: everything after the stripped leading ctx slot.
    private final Type[] userTypes; // declared (generic) types, length == user param count
    private final boolean varargs; // whether the last user parameter is a Java varargs
    private final Class<?> varargComponent; // element type of that varargs, else null
    private final Type returnType;

    // Construction is from a raw Method: internal plumbing the ref-based entrypoints
    // ({@code Resonate}/{@code Context} run/rpc/register) resolve into. Package-private so the
    // public SDK surface is uniformly ref-based, never a raw Method.

    /** Build from a static (or already-target-bound) {@link Method}. */
    Durable(Method fn) {
        this(null, fn);
    }

    /**
     * Build from a {@code Method} and the instance it is invoked on ({@code null} for a static
     * method). The instance is the analogue of a Python bound method / callable instance — {@code
     * self} is already supplied, so the function's first user-visible parameter is still {@code
     * ctx}.
     */
    Durable(Object target, Method fn) {
        if (fn == null) {
            throw new ApplicationError("expected a callable, got null");
        }
        this.target = target;
        this.fn = fn;
        this.name = fn.getName();

        int paramCount = fn.getParameterCount();
        if (paramCount == 0) {
            throw new ApplicationError(name + ": durable function must accept a Context as its first argument");
        }
        // ``ctx`` is injected positionally and stripped by position, so the first parameter must be
        // able to hold it. In Java the only shape that cannot is a lone varargs parameter — the
        // tell-tale of an unwrapped (*args)-style wrapper — since varargs is always the last
        // parameter. Reject it with a clear message rather than failing confusingly later.
        if (fn.isVarArgs() && paramCount == 1) {
            throw new ApplicationError(name
                    + ": durable function must accept a Context as its first positional argument, but its"
                    + " only parameter is variadic");
        }

        Type[] generic = fn.getGenericParameterTypes();
        Class<?>[] raw = fn.getParameterTypes();
        int userCount = paramCount - 1;
        this.userTypes = new Type[userCount];
        System.arraycopy(generic, 1, this.userTypes, 0, userCount);
        this.varargs = fn.isVarArgs();
        this.varargComponent = varargs ? raw[paramCount - 1].getComponentType() : null;
        this.returnType = fn.getGenericReturnType();

        fn.setAccessible(true);
    }

    /** Source name (child context {@code func_name} + error messages). */
    public String name() {
        return name;
    }

    /**
     * Validate a call's arity and pack it into a serializable {@link Args}.
     *
     * <p>Called at dispatch ({@code ctx.run}) time. The injected {@code Context} is never part of
     * the payload. Java has no keyword arguments, so only positional {@code args} are packed; the
     * {@code kwargs} slot stays empty. Raises {@link ApplicationError} if the call does not match
     * the signature.
     */
    public Args packArgs(Object... args) {
        checkArity(args == null ? 0 : args.length);
        List<Object> list = new ArrayList<>();
        if (args != null) {
            for (Object a : args) {
                list.add(a);
            }
        }
        return new Args(List.copyOf(list), Map.of());
    }

    /**
     * Re-bind {@code packed} to the signature and execute the function.
     *
     * <p>When {@code coerceArgs} is {@code true} (the root / dispatched path) arguments are coerced
     * to their declared types: that path's args arrive as JSON builtins decoded from the persisted
     * param, so they must be reshaped. A local {@code ctx.run} child passes {@code false}: its
     * arguments are never serialized into the param, so the in-memory objects reach the function
     * verbatim. Arity is validated regardless.
     *
     * <p>{@code ctx} is injected as the first positional argument. The function's return value is
     * returned as-is.
     */
    public Object invoke(Object ctx, Args packed, boolean coerceArgs) {
        List<Object> provided = packed.args();
        checkArity(provided.size());

        Object[] call = bind(ctx, provided, coerceArgs);
        try {
            return fn.invoke(target, call);
        } catch (InvocationTargetException exc) {
            // Surface the user's exception unchanged (ApplicationError, etc.), mirroring Python's
            // bare propagation of a raising function.
            Throwable cause = exc.getCause();
            if (cause instanceof RuntimeException re) {
                throw re;
            }
            if (cause instanceof Error err) {
                throw err;
            }
            throw new ApplicationError(name + ": " + (cause == null ? exc.getMessage() : cause.getMessage()));
        } catch (IllegalAccessException exc) {
            throw new ApplicationError(name + ": " + exc.getMessage());
        }
    }

    /** {@link #invoke(Object, Args, boolean)} with {@code coerceArgs = true} (the dispatched path). */
    public Object invoke(Object ctx, Args packed) {
        return invoke(ctx, packed, true);
    }

    /**
     * Reshape a recovered return value into the declared return type — the symmetric counterpart of
     * argument coercion. A pass-through return type ({@code Object} / {@code void} / {@code Void})
     * reshapes nothing.
     */
    public Object coerceResult(Object value) {
        return coerceValue(value, returnType, null);
    }

    /**
     * The declared return type as a value safe to hand to a coercion call. A pass-through annotation
     * collapses to {@link Object}, so a top-level decode against this type is the same no-op {@link
     * #coerceResult} performs for those cases.
     */
    public Type returnType() {
        return isPassthrough(returnType) ? Object.class : returnType;
    }

    // --- internals ---

    private void checkArity(int n) {
        int fixed = varargs ? userTypes.length - 1 : userTypes.length;
        boolean ok = varargs ? n >= fixed : n == fixed;
        if (!ok) {
            String shape = varargs ? "at least " + fixed : Integer.toString(fixed);
            throw new ApplicationError(name + ": expected " + shape + " positional argument(s), got " + n);
        }
    }

    /** Build the reflective argument array {@code [ctx, user-args...]}, coercing when asked. */
    private Object[] bind(Object ctx, List<Object> provided, boolean coerce) {
        int fixed = varargs ? userTypes.length - 1 : userTypes.length;
        // ctx + fixed user params (+ one collapsed varargs array if present)
        Object[] call = new Object[1 + fixed + (varargs ? 1 : 0)];
        call[0] = ctx;

        for (int i = 0; i < fixed; i++) {
            Object v = provided.get(i);
            call[1 + i] = coerce ? coerceValue(v, userTypes[i], name) : v;
        }

        if (varargs) {
            int tailLen = provided.size() - fixed;
            Object array = java.lang.reflect.Array.newInstance(varargComponent, tailLen);
            for (int i = 0; i < tailLen; i++) {
                Object v = provided.get(fixed + i);
                java.lang.reflect.Array.set(array, i, coerce ? coerceValue(v, varargComponent, name) : v);
            }
            call[call.length - 1] = array;
        }
        return call;
    }

    /**
     * Reshape one value to {@code type} via Jackson, or pass it through. A {@code type} that carries
     * nothing to reshape (see {@link #isPassthrough}) leaves the value untouched. {@code context},
     * when non-null, is the function name added to a failure's message (the analogue of Python's
     * {@code add_note}).
     */
    private static Object coerceValue(Object value, Type type, String context) {
        if (isPassthrough(type)) {
            return value;
        }
        try {
            return MAPPER.convertValue(value, MAPPER.getTypeFactory().constructType(type));
        } catch (IllegalArgumentException exc) {
            if (context == null) {
                throw new SerializationError(exc);
            }
            throw new SerializationError(new IllegalArgumentException(
                    exc.getMessage() + " (while binding arguments for " + context + ")", exc));
        }
    }

    /**
     * Whether {@code type} carries no information to reshape a value. The Java analogues of Python's
     * four pass-through cases: {@code Object} (the {@code Any} analogue), {@code void} / {@code
     * Void} (the {@code None} / {@code -> None} analogue), and {@code null} (a missing type).
     */
    private static boolean isPassthrough(Type type) {
        return type == null || type == Object.class || type == void.class || type == Void.class;
    }
}
